aop/src/main/java/io/micronaut/aop/util/KotlinInvocationContextUtils.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.aop.util;
+
+import io.micronaut.aop.Interceptor;
+import io.micronaut.aop.MethodInvocationContext;
+import io.micronaut.core.annotation.Experimental;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.type.MutableArgumentValue;
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.core.util.KotlinUtils;
+import kotlin.coroutines.Continuation;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Consumer;
+
+/**
+ * <p>Internal Utility methods for working with Kotlin <code>suspend</code> functions with {@link io.micronaut.aop.MethodInterceptor}</p>.
+ *
+ * @author Denis stepanov
+ * @since 2.1.0
+ */
+@Internal
+@Experimental
+public class KotlinInvocationContextUtils {
+
+    /**
+     * Checks if the method invocation is a Kotlin coroutine.
+     *
+     * @param context {@link MethodInvocationContext}
+     * @return true if Kotlin coroutine
+     */
+    public static boolean isKotlinCoroutine(MethodInvocationContext<Object, Object> context) {
+        if (!KotlinUtils.KOTLIN_COROUTINES_SUPPORTED || !context.getExecutableMethod().isSuspend()) {
+            return false;
+        }
+        Collection<MutableArgumentValue<?>> arguments = context.getParameters().values();
```

Comment:
Think it would be more efficient to use `getParameterValues()` and check the array since this will not build the `MutableArgumentValue` objects and there is no requirement here to use the `MutableArgumentValue` interface